bee-network/bee-autopeering/src/peering/manager.rs

```diff
@@ -0,0 +1,774 @@
+            status = false;
+
+            // Fire `OutgoingPeering` event with status = `false`.
+            publish_peering_event::<OUTGOING>(
+                peer.clone(),
+                status,
+                ctx.local,
+                ctx.event_tx,
+                ctx.inbound_nbh,
+                ctx.outbound_nbh,
+            );
+
+            // Drop that peer.
+            send_drop_peering_request_to_peer(peer, ctx.server_tx, ctx.event_tx, ctx.inbound_nbh, ctx.outbound_nbh);
+        } else if ctx.outbound_nbh.insert_neighbor(peer.clone(), ctx.local) {
+            // Update the neighbor filter.
+            nb_filter.add(*peer.peer_id());
+
+            // Fire `OutgoingPeering` event with status = `true`.
+            publish_peering_event::<OUTGOING>(peer, status, ctx.local, ctx.event_tx, ctx.inbound_nbh, ctx.outbound_nbh);
+        } else {
+            log::debug!("Failed to add neighbor to outbound neighborhood after successful peering request");
+        }
+    } else {
+        log::debug!("Peering by {} denied.", ctx.peer_id);
+    }
+
+    // Send the response notification.
+    if let Some(tx) = peer_reqval.response_tx {
+        tx.send(ctx.msg_bytes.to_vec()).expect("error sending response signal");
+    }
+}
+
+fn handle_drop_request<V: NeighborValidator>(
+    _drop_req: DropPeeringRequest,
+    ctx: RecvContext,
+    nb_filter: &NeighborFilter<V>,
+) {
+    log::trace!("Handling drop request.");
+
+    let mut removed_nb = ctx.inbound_nbh.remove_neighbor(ctx.peer_id);
+
+    if let Some(nb) = ctx.outbound_nbh.remove_neighbor(ctx.peer_id) {
+        removed_nb.replace(nb);
+
+        nb_filter.add(*ctx.peer_id);
+
+        // TODO: trigger immediate outbound neighborhood update; currently we wait for the next interval
+    }
+
+    if removed_nb.is_some() {
+        send_drop_peering_request_to_addr(
+            ctx.peer_addr,
+            *ctx.peer_id,
+            ctx.server_tx,
+            ctx.event_tx,
+            ctx.inbound_nbh,
+            ctx.outbound_nbh,
+        );
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+// SENDING
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/// Initiates a peering request.
+///
+/// This function is blocking, but at most for `RESPONSE_TIMEOUT` seconds.
+pub(crate) async fn begin_peering(
+    peer_id: &PeerId,
+    active_peers: &ActivePeersList,
+    request_mngr: &RequestManager,
+    server_tx: &ServerTx,
+    local: &Local,
+) -> Option<Status> {
+    let (response_tx, response_rx) = request::response_chan();
+
+    send_peering_request_to_peer(peer_id, active_peers, request_mngr, server_tx, Some(response_tx), local);
+
+    match tokio::time::timeout(RESPONSE_TIMEOUT, response_rx).await {
+        Ok(Ok(bytes)) => {
+            let peer_res = PeeringResponse::from_protobuf(&bytes).expect("error decoding peering response");
```

Comment:
For this and following, some `expect` if you want to handle them

Reply:
oh, yea. Handled!